test/test_nn.py

```diff
@@ -8572,5 +8607,109 @@ def _buildEquivalentAffineTransforms3d(device, input_size, output_size, angle_ra
 # end TestNN.test_affine_* helpers
 
 
+# The following are some helpers for TestNN.test_transformer_number_match
+class DataBatch:
+    "Object for holding a batch of data with mask during training."
+    def __init__(self, src, trg=None, pad=0):
+        self.src = src.transpose(0, 1).contiguous()
+        self.src_mask = (src != pad).unsqueeze(-2)
+        if trg is not None:
+            self.trg = trg[:, :-1].transpose(0, 1).contiguous()
+            self.trg_y = trg[:, 1:].transpose(0, 1).contiguous()
+            self.ntokens = (self.trg_y != pad).data.sum().item()
+
+class NoamOpt:
```

Comment:
There should be an [optimizer](https://pytorch.org/docs/stable/optim.html) that already does this and if not, if this is common, we should add it explicitly. What features do our current optimizers lack, that this one provides?